lib/OA/Admin/PasswordRecovery.php

```diff
@@ -217,37 +217,66 @@ function checkAccess()
     function sendRecoveryEmail($email)
     {
         $aConf = &$GLOBALS['_MAX']['CONF'];
-        $aPref = $GLOBALS['_MAX']['PREF'];
+        $sent = 0;
 
+        // Find all users matching the specified email address -
+        // the email address may be associated with multiple users
         $aUsers = $this->_dal->searchMatchingUsers($email);
 
-        $aEmails = array();
+        // Send a separate password reset link in an email for each
+        // of the users found that match the email address
         foreach ($aUsers as $u) {
-            $aEmails[$u['email_address']][] = $u;
-        }
-
-        $sent = 0;
-        foreach ($aEmails as $email => $aUsers) {
-            $text = '';
-            foreach ($aUsers as $u) {
-                $recoveryId = $this->_dal->generateRecoveryId($u['user_id']);
-
-                $header = $GLOBALS['strUser']." {$u['contact_name']}";
-                $text .= $header."\n".str_repeat('-', strlen($header))."\n";
-                $text .= $GLOBALS['strUsername'].": {$u['username']}\n";
-                $text .= $GLOBALS['strPwdRecResetLink'].": ";
-                $text .= Max::constructURL(MAX_URL_ADMIN, "password-recovery.php?id={$recoveryId}")."\n\n";
+            // Generate the password reset email subject
+            $emailSubject = sprintf($GLOBALS['strPwdRecEmailPwdRecovery'], PRODUCT_NAME);
+
+            // Set the name and email address that the password reset email
+            // is to be sent from. Ideally, this will be the administrative
+            // account of the Revive Adserver installation - but fall back
+            // to the user's own email address if required
+            if (!empty($aConf['email']['fromName'])) {
+                $emailFromName = $aConf['email']['fromName'];
+            } else {
+                $emailFromName = $email;
+            }
+            if (!empty($aConf['email']['fromAddress'])) {
+                // Use the administrative account details
+                $emailFromAddress = $aConf['email']['fromAddress'];
+            } else {
+                $emailFromAddress = $email;
             }
 
-            // Hack
-            $aConf['email']['admin_name'] = $aPref['admin_fullname'];
```

Comment:
What about this "hack"? I suppose it was there for a reason?

Reply:
I think this hack was supposed to be there to get the admin details of the system, to be used as who the email should be sent from.

However, the loading of the $aPref array simply didn't work at all - the password recovery system runs while there's no user logged in, so there were no user preferences available in that array.

I replaced this with the details of the system administrator name and email that can be set in the installation's configuration file, as this is definitely available when there's no one logged in, and makes the most sense to me for using in this context.